File: load.js
// Connect to your MongoDB instance
mongosh

// Create a test database and collection with sample data
use loadtest
for(let i = 0; i < 100000; i++) {
    db.testCollection.insertOne({
        index: i,
        data: "x".repeat(1000),
        timestamp: new Date(),
        nested: {
            field1: Math.random() * 1000,
            field2: "sample_" + i,
            array: Array(50).fill(0).map(() => Math.random())
        }
    });
}



// Run expensive aggregation pipelines
db.testCollection.aggregate([
    { $match: { index: { $gt: 0 } } },
    { $sort: { "nested.field1": -1 } },
    { $group: {
        _id: { $mod: ["$index", 100] },
        count: { $sum: 1 },
        avgField: { $avg: "$nested.field1" },
        data: { $push: "$$ROOT" }
    }},
    { $sort: { count: -1 } }
]);

// Run without indexes to stress CPU
db.testCollection.find({ "nested.field1": { $gt: 500 } }).sort({ index: -1 });

// Large document operations
db.testCollection.aggregate([
    { $lookup: {
        from: "testCollection",
        localField: "index",
        foreignField: "nested.field1",
        as: "joined"
    }},
    { $unwind: "$joined" }
]);

// Load large result sets into memory
db.testCollection.find({}).toArray();

File: load.py
from pymongo import MongoClient
from concurrent.futures import ThreadPoolExecutor
import random
import time

client = MongoClient('mongodb://localhost:27017/')
db = client.loadtest

def generate_load(thread_id):
    """Generate continuous load"""
    while True:
        # Random operations
        operation = random.choice(['insert', 'query', 'aggregate', 'update'])
        
        if operation == 'insert':
            db.testCollection.insert_many([
                {
                    'thread': thread_id,
                    'data': 'x' * 5000,
                    'value': random.randint(1, 1000),
                    'timestamp': time.time()
                } for _ in range(100)
            ])
        
        elif operation == 'query':
            list(db.testCollection.find({'value': {'$gt': random.randint(1, 500)}}).limit(1000))
        
        elif operation == 'aggregate':
            list(db.testCollection.aggregate([
                {'$match': {'value': {'$gt': 400}}},
                {'$group': {'_id': '$thread', 'count': {'$sum': 1}}},
                {'$sort': {'count': -1}}
            ]))
        
        elif operation == 'update':
            db.testCollection.update_many(
                {'value': {'$lt': 100}},
                {'$set': {'updated': time.time()}}
            )

# Run with multiple threads
with ThreadPoolExecutor(max_workers=8) as executor:
    for i in range(8):
        executor.submit(generate_load, i)
